### PR TITLE
Override subscriptionManager.middleware.destroy to also remove sub entry

### PR DESCRIPTION
**Pull request by jiexi**
## Explanation

Ensures that the subscription entry for a subscription is removed when the middleware is destroyed

## References

Related: https://github.com/MetaMask/metamask-extension/pull/28751

### `@metamask/package-a`

- **<CATEGORY>**: Your change here
- **<CATEGORY>**: Your change here

### `@metamask/package-b`

- **<CATEGORY>**: Your change here
- **<CATEGORY>**: Your change here

## Checklist

- [ ] I've updated the test suite for new or updated code as appropriate
- [ ] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [ ] I've highlighted breaking changes using the "BREAKING" category above as appropriate
- [ ] I've prepared draft pull requests for clients and consumer packages to resolve any breaking changes

### CODE REVIEW

**Comment by jiexi** [comment]
@metamaskbot publish-preview

**Comment by jiexi** [comment]
@metamaskbot publish-previews

**Comment by jiexi** [comment]
@metamaskbot publish-previews

**Comment by jiexi** [comment]
@metamaskbot publish-previews

**Comment by jiexi** [comment]
@metamaskbot publish-previews

**Comment by jiexi** [comment]
@metamaskbot publish-previews

**Comment by jiexi** [comment]
@metamaskbot publish-previews

**Comment by jiexi** [comment]
@metamaskbot publish-previews

**Comment by jiexi** [comment]
@metamaskbot publish-previews

**Comment by jiexi** [comment]
@metamaskbot publish-previews

**Comment by jiexi** [comment]
@metamaskbot publish-previews

File: packages/multichain/src/middlewares/MultichainSubscriptionManager.test.ts [hunk 1]
```diff
@@ -151,9 +153,14 @@ describe('MultichainSubscriptionManager', () => {
     multichainSubscriptionManager.subscribe({ scope, origin, tabId });
     multichainSubscriptionManager.unsubscribeByOriginAndTabId(origin, tabId);
 
-    mockSubscriptionManager.events.on.mock.calls[0][1](
-      newHeadsNotificationMock,
-    );
+    expect(mockSubscriptionManager.destroy).toHaveBeenCalled();
+  });
+
+  it('should unsubscribe when the middleware is destroyed', () => {
```
Comment (adonesky1) [inline packages/multichain/src/middlewares/MultichainSubscriptionManager.test.ts:159]: Should there be an assertion in this test about the state of subscriptions? Or I suppose you can't access this 🤔

File: packages/multichain/src/middlewares/MultichainSubscriptionManager.ts [hunk 2]
```diff
@@ -108,10 +112,17 @@ export class MultichainSubscriptionManager extends SafeEventEmitter {
...
+    const newSubscriptionManagerEntry = {
       ...subscriptionKey,
       subscriptionManager,
-    });
+    };
+    subscriptionManager.destroy = subscriptionManager.middleware.destroy;
+    subscriptionManager.middleware.destroy = this.#unsubscribe.bind(
+      this,
+      newSubscriptionManagerEntry,
+    );
```
Comment (adonesky1) [inline packages/multichain/src/middlewares/MultichainSubscriptionManager.ts:123]: 🤔 maybe I'm missing it but any reason we could do this instead?
```suggestion
    subscriptionManager.destroy = this.#unsubscribe.bind(
      this,
      newSubscriptionManagerEntry,
    );
```

File: packages/multichain/src/middlewares/MultichainSubscriptionManager.test.ts (hunk 1 above)
Comment (jiexi) [reply packages/multichain/src/middlewares/MultichainSubscriptionManager.test.ts:159]: Like you said, it's private so you can't actually check the internal state

File: packages/multichain/src/middlewares/MultichainSubscriptionManager.ts (hunk 2 above)
Comment (jiexi) [reply packages/multichain/src/middlewares/MultichainSubscriptionManager.ts:123]: https://github.com/MetaMask/eth-json-rpc-filters/blob/main/subscriptionManager.js#L24

There's two issues: 

* `SubscriptionManager.middleware.destroy` is set to the `SubscriptionManager.destroy` function (as opposed to being a function that calls `SubscriptionManager.destroy`)
* `SubscriptionManager.destroy` is not actually exported/exposed and only accessible through `SubscriptionManager.middleware.destroy`

Comment (adonesky1) [reply packages/multichain/src/middlewares/MultichainSubscriptionManager.ts:123]: oh wow I totally misread that

**Review by adonesky1 — APPROVED** [review]
LGTM

**Pull request merged**